// tefcha/dist/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mergeDefaultConfig = exports.defaultConfig = exports.Flowchart = exports.createFlowchart = exports.Renderer = exports.render = void 0;
var flowchart_1 = require("./flowchart");
Object.defineProperty(exports, "createFlowchart", { enumerable: true, get: function () { return flowchart_1.createFlowchart; } });
Object.defineProperty(exports, "Flowchart", { enumerable: true, get: function () { return flowchart_1.Flowchart; } });
var config_1 = require("./config");
Object.defineProperty(exports, "defaultConfig", { enumerable: true, get: function () { return config_1.defaultConfig; } });
Object.defineProperty(exports, "mergeDefaultConfig", { enumerable: true, get: function () { return config_1.mergeDefaultConfig; } });
var svg_1 = require("./renderer/svg");
Object.defineProperty(exports, "render", { enumerable: true, get: function () { return svg_1.render; } });
Object.defineProperty(exports, "Renderer", { enumerable: true, get: function () { return svg_1.Renderer; } });
